File: solutions/004/004.cc
#include <iostream>

bool is_palindrome(int orig) {
    int rev = 0, n = orig;
    while (n  > 0) {
        rev = rev * 10 + n % 10;
        n /= 10;
    }
    return orig == rev;
}

int main() {
    int largest = 0, last = 999, first = 100;
    for (int a = last; a >= first; a--)
        for (int b = last; b >= first; b--) {
            int calculated = a * b;
            if (calculated > largest && is_palindrome(calculated))
                largest = calculated;
        }
    std::cout << largest << std::endl;
}

File: solutions/004/004.rs
fn is_palindrome(orig : i32) -> bool {
    let mut rev = 0;
    let mut n = orig;
    while n > 0 {
        rev = rev * 10 + n % 10;
        n /= 10;
    }
    return orig == rev;
}

fn main() {
    let mut largest = 0;
    let last = 999;
    let first = 100;

    let mut a = last;
    while a >= first {
        let mut b = last;
        while b >= first {
            let calculated = a * b;
            if calculated > largest && is_palindrome(calculated) {
                largest = calculated;
            }
            b -= 1;
        }
        a -= 1;
    }
    println!("{}", largest);
}
